### PR TITLE
Planet attachments

### PR DESCRIPTION
**Pull request by ViddeM**
Also adds planet names and some other stuff.

### CODE REVIEW

File: Roadmap.md [hunk 1]
```diff
@@ -84,6 +85,22 @@
      - ✅ council keleres faction
  - ✅ Strategy cards (speaker handling etc)
  - ❌ Real names for things (currently we only use the ID for e.g objectives)
```
Comment (ViddeM) [inline Roadmap.md:87]: Dessa är TODO mest för att komma ihåg att kolla på dem, flera av dessa är klara (åtminstonde i BE)

File: backend/game/src/gameplay/game_event_handler.rs [hunk 2]
```diff
@@ -932,25 +1016,106 @@ pub fn update_game_state(
         Event::SetPlanetOwner { player, planet } => {
             game_state.assert_expansion(&planet.info().expansion)?;
 
-            // Give the planet to the new owner.
             if let Some(p) = &player {
+                ensure!(game_state.players.contains_key(p), "Player does not exist");
+
+                let (_, attachments) = game_state.players.iter_mut().fold(
```
Comment (hulthe) [inline backend/game/src/gameplay/game_event_handler.rs:1022]: find_map

File: backend/game/src/gameplay/game_event_handler.rs [hunk 3]
```diff
@@ -932,25 +1016,106 @@ pub fn update_game_state(
...
-                    bail!("Player does not exist?")
+                    bail!("Player does not exist? This is a bug!")
                 };
-                player.planets.insert(planet.clone());
+                let planet_info = planet.info();
+                player.planets.insert(
+                    planet,
+                    attachments
+                        .into_iter()
+                        .map(|a| a.match_planet(&planet_info))
```
Comment (hulthe) [inline backend/game/src/gameplay/game_event_handler.rs:1040]: same question about match_planet

**Review by hulthe — COMMENTED** [review]
Left some comments. Also, what is your opinion on clippy?  Personally, I compulsively use it, and like to make sure that all clippy lints are fixed, (or silenced if I disagree with them).

I'm asking because I noticed we have a bunch of clippy warnings now, although not a problem for this PR since it only introduces 1 or 2 new ones.

File: backend/game/src/gameplay/agenda.rs [hunk 4]
```diff
@@ -104,16 +105,22 @@ impl VoteState {
...
-                        .values()
-                        .any(|player| player.planets.contains(planet))
+                // Filter to planets that are held by a player & map to a tuple of the planet, planet attachments.
+                .filter_map(|p| {
+                    game.players.values().fold(None, |acc, player| {
+                        if player.planets.contains_key(&p) {
+                            return Some((p.clone(), &player.planets[&p]));
+                        }
+                        acc
+                    })
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/agenda.rs:117]: Very nice

File: backend/game/src/gameplay/event.rs [hunk 5]
```diff
@@ -329,6 +340,26 @@ pub enum Event {
...
+        planet: Planet,
+        /// The attachment that is added.
+        attachment: PlanetAttachment,
+    },
+
+    /// Remove a planet attachment from a planet.
+    RemovePlanetAttachment {
+        /// The player who owns the planet.
+        player: PlayerId,
+        /// The planet that should receive the attachment.
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/event.rs:358]: Right, copy paste 😅

File: backend/game/src/gameplay/event.rs [hunk 6]
```diff
@@ -329,6 +340,26 @@ pub enum Event {
...
+        planet: Planet,
+        /// The attachment that is added.
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/event.rs]: Done

File: backend/game/src/gameplay/game_event_handler.rs [hunk 7]
```diff
@@ -297,14 +299,36 @@ pub fn update_game_state(
             let Some(current_player_id) = &game_state.current_player else {
                 bail!("no current player");
             };
+            ensure!(
+                game_state.players.contains_key(current_player_id),
+                "Current player doesn't exist?"
+            );
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/game_event_handler.rs]: Fair enogh x)

removed.

File: backend/game/src/gameplay/game_event_handler.rs [hunk 8]
```diff
@@ -321,7 +345,67 @@ pub fn update_game_state(
...
+                );
+
+                ensure!(
+                    !tactical.planet_attachments.contains_key(&planet),
+                    "Planet has already been explored this turn!"
+                );
+
+                let attachment_info = attachment.info();
+                let Some(attachment_planet_trait) = &attachment_info.planet_trait else {
+                    // TODO: Somewhat of an exploit of the fact that currently, only planets with planet_trait requirements are exploration attachments (and all of them have it).
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/game_event_handler.rs]: Yeah, check if it makes more sense now?

File: backend/game/src/gameplay/game_event_handler.rs [hunk 9]
```diff
@@ -297,14 +299,36 @@ pub fn update_game_state(
...
+            let (current_owner, attachments) =
+                game_state
+                    .players
+                    .iter_mut()
+                    .fold((None, HashSet::new()), |acc, (id, player)| {
+                        if let Some(attachments) = player.planets.remove(&planet) {
+                            return (Some(id.clone()), attachments);
+                        }
+                        acc
+                    });
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/game_event_handler.rs]: Not sure if it became much better but perhaps a little more readable.

File: backend/game/src/gameplay/game_event_handler.rs [hunk 10]
```diff
@@ -297,14 +299,36 @@ pub fn update_game_state(
...
 
             let planet_system = System::for_planet(&planet)?;
 
-            current_player.planets.insert(planet.clone());
+            let planet_info = planet.info();
+            current_player.planets.insert(
+                planet.clone(),
+                attachments
+                    .into_iter()
+                    .map(|a| a.match_planet(&planet_info))
```
[the remark this answers is not in the document]
Comment (ViddeM) [reply backend/game/src/gameplay/game_event_handler.rs:325]: No we do not,

File: backend/game/src/gameplay/game_event_handler.rs (hunk 3 above)
Comment (ViddeM) [reply backend/game/src/gameplay/game_event_handler.rs:1038]: Yep good point

**Pull request merged**